ats_fastcgi experimental plugin

oschaaf opened the pull request:
As presented at the summit in Cork. 
See `README.md` for more information.

oschaaf commented:
The RAT complaints are on trivial files in `examples/`, do we need to add the asf license to them regardless, or can we ignore those?

There's one complaint that needs a closer look, by someone more knowledgeable then me: `src/configuru.hpp`

zwoop commented:
My preference would be that you add the license blurb where possible. I.e. if a config file can not handle comments, then you can add that particular file to the RAT exclusion, but in general, it's just a lot better to minimize the exclusions when at all possible.

oschaaf commented:
RAT & clang passing, but now autest complains, at a glance on seemingly unrelated stuff

oschaaf commented:
Some more notes:
- I tested this with ATS 7.1.3
- Running ASP.net via mono fastcgi [1], it seems something is not right when the connection pooling feature is enabled, because after a couple of requests things the browser will hang. This is probably because of some internal state getting messed up due to a missed edge case. Disabling connection pooling by setting `proxy.config.http.fcgi.host.max_requests` to `0` resolves it for me.

Samples of .NET handling requests:
http://104.199.7.245/test.aspx
http://104.199.7.245/test.css

oschaaf commented:
(resolved small conflict in rat excludes that arose by a commit on master)

bryancall commented on plugins/experimental/fastcgi/src/ats_fcgi_client.cc:
```diff
@@ -0,0 +1,646 @@
+    Headers &h               = transaction.getClientRequest().getHeaders();
+
+    if (h.isInitialized()) {
+      string key("Content-Length");
```
What if the POST is chunked?

oschaaf replied:
Hmm, searching for 'fastcgi chunked post data' makes it look like it is at least common for FastCGI implementations not to take that well. (top result: https://www.jeffgeerling.com/blog/2017/apache-fastcgi-proxyfcgi-and-empty-post-bodies-chunked-transfer)
Having said that, maybe we should explicitly handle the case and respond with an error when someone tries to POST chunked data, and add a TODO in the code to support it?

oschaaf replied:
Testing it out:

```
oschaaf@ats-fastcgi:~/weamp/newats⟫ curl -v -i -d @"test" -H "Transfer-Encoding: chunked"  127.0.0.1/test.aspx
Warning: Couldn't read data from file "test", this makes an empty POST.
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
> POST /test.aspx HTTP/1.1
> Host: 127.0.0.1
> User-Agent: curl/7.47.0
> Accept: */*
> Transfer-Encoding: chunked
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 5 out of 0 bytes
< HTTP/1.1 500 Server Error
HTTP/1.1 500 Server Error
< Age: 0
Age: 0
< Date: Thu, 05 Jul 2018 21:17:11 GMT
Date: Thu, 05 Jul 2018 21:17:11 GMT
< Transfer-Encoding: chunked
Transfer-Encoding: chunked
< Connection: keep-alive
Connection: keep-alive
< Server: ATS/9.0.0
Server: ATS/9.0.0

< 
* Connection #0 to host 127.0.0.1 left intact
Content length parameter not an integer.
```

oschaaf replied:
(sorry, bad example, but also in the case where I actually do send an *existing*, handling is the same)

bryancall commented on plugins/experimental/fastcgi/src/ats_fcgi_client.cc:
```diff
@@ -0,0 +1,646 @@
+  Headers &h               = transaction.getClientRequest().getHeaders();
+  if (h.isInitialized()) {
+    atscppapi::header_field_iterator it = h.begin();
+    for (it = h.begin(); it != h.end(); ++it) {
```
Better to do a range-based for loops, or clang-tidy will change it

bryancall commented on plugins/experimental/fastcgi/src/request_queue.h:
```diff
@@ -0,0 +1,45 @@
+ * limitations under the License.
+ */
+
+#ifndef REQUEST_QUEUE_H
```
We switch to use pragma once

oschaaf commented:
@bryancall @SolidWallOfCode thanks for the feedback.
(Sorry for the late reply, I was OOO. I will address the review comments soon)

oschaaf commented:
So, I set up this plugin on a development machine based on the master branch, for addressing the comments in here. Previously, on 7.x it worked, but on master the plugin hangs after the intercept is hooked. 
```
[Jun 27 15:16:00.557] {0x7f63d5b90700} DEBUG: <HttpSM.cc:1957 (state_send_server_request_header)> (http) [0] [&HttpSM::state_send_server_request_header, VC_EVENT_WRITE_COMPLETE]
[Jun 27 15:16:30.590] {0x7f63d5b90700} DEBUG: <HttpSM.cc:2529 (main_handler)> (http) [0] [HttpSM::main_handler, VC_EVENT_INACTIVITY_TIMEOUT]
```

This requires some digging into first, before I can go on and address comments.

oschaaf commented:
Once https://github.com/apache/trafficserver/pull/3927 lands, I plan to rebase this and proceed to addressing comments

oschaaf commented:
I think I addressed the comments in here, PTAL?

oschaaf commented:
@bryancall I addressed your comment, I think, WDYT?

oschaaf commented:
@bryancall ping :-)

bryancall commented:
@oschaaf looks good, can please squash the commits

oschaaf commented:
@bryancall  Sure, I will squash them

oschaaf commented:
@bryancall Rebased onto the head of master, squashed & force pushed.

bryancall commented:
Cherry picked to 8.1.0
